### DISCUSSION TITLE
Hill elasto-plasticity x PFF

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#27746] [Q&A Modules: Solid mechanics]
Hello, I want to add damage to the hill plasticity (HillElastoPlasticityStressUpdate) through the PF nonconserved action. From what I understand I need to define the new degraded stress and its derivatives and the free energy and its derivatives (dstress_dc, F, dFdc, d2Fdc2,  d2Fdcdstrain).  I am getting errors when declaring the derivatives in the constructor  ex:
```
error: 'declarePropertyDerivative' was not declared in this scope; did you mean 'declarePropertyByName'?
   32 |     _dstress_dc(declarePropertyDerivative<RankTwoTensor>(_base_name + "stress", coupledName("c", 0))),
      |                 ^~~~~~~~~~~~~~~~~~~~~~~~~
      |                 declarePropertyByName
```
 
though I did include the DerivativeMaterialInterface class. I also don't see in which of the inherited class is the elastic strain called: I see that HillElastoPlas. inherits from the RadialReturnStressUpdate (https://mooseframework.inl.gov/source/materials/RadialReturnStressUpdate.html) which should have _elastic_strain_old[_qp] but it's also not declared with the error showing as : `error: '_elastic_strain_old' was not declared in this scope;`.

**Comment by GiudGiud** [comment]
You may have to use the `using` syntax to see the ` _elastic_strain_old` attribute from the parent class

> though I did include the DerivativeMaterialInterface class

what does the class definition look like?

**Reply by ghost** [reply]
I figured it out, thanks ! Now my question is regarding how the HillElastoPlasticityStressUpdate and the ComputeMultipleInelasticStress are correlated. I see from the diagram in https://mooseframework.inl.gov/source/materials/ComputeMultipleInelasticStress.html that the trial irreversible stress is computed in HillElastoPlasticityStressUpdate, so this is the stress_new in HillElastoPlasticityStressUpdate ? and the final stress is the stress in ComputeFiniteStrainElasticStress ?

**Comment by ghost** [comment]
Can we use both the HillElastoPlasticityStressUpdate and ComputeCrackedStress together? I keep getting error messages, is it due to a wrong input file or are we not supposed to use them together? Thanks!

**Reply by ghost** [reply]
```*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'rotation_increment', requested by 'hill_tensor' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_face' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_neighbor' is not defined on block 0
```
For the input file as :
```[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 3
  []
  [./BL]
    type = BoundingBoxNodeSetGenerator
    new_boundary = BL
    input = gen
    top_right = '1 0 0'
    bottom_left = '0 0 0'
  [../]
  [./BR]
    type = BoundingBoxNodeSetGenerator
    new_boundary = BR
    input = BL
    top_right = '1 0 1'
    bottom_left = '1 0 0'
  [../]
[]


[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Variables]
  [./disp_x]
  [../]
  [./disp_y]
  [../]
  [./disp_z]
  [../]
[]

[AuxVariables]
  [./resid_x]
  [../]
  [./resid_y]
  [../]
  [./resid_z]
  [../]
[]

[Modules]
[./TensorMechanics]
    [./Master]
      [./All]
        add_variables = true
        strain = FINITE
        save_in = 'resid_x resid_y resid_z'
       strain_base_name = uncracked
        incremental = true
      [../]
    [../]
[../]
  [./PhaseField]
    [./Nonconserved]
      [./c]
        free_energy = E_el
        mobility = L
        kappa = kappa_op
      [../]
    [../]
  [../]
[]  

[Variables]
  [./c]
  [../]
[]

[AuxVariables]
  [./bounds_dummy]
  [../]
[]

[BCs]
  [./bottomy]
    type = DirichletBC
    variable = disp_y
    boundary =bottom
    value = 0.0
  [../]
  [./topy]
    type = FunctionDirichletBC
    variable = disp_y
    boundary = top
    function ='t'
  [../]
  [./xfix]
    type = DirichletBC
    variable = disp_x
    boundary = 'BR'
    value = 0
  [../]
    [./zfix]
    type = DirichletBC
    variable = disp_z
    boundary = 'BL'
    value = 0
  [../]
[]

[Materials]
   [./cracked_stress]
    type = ComputeCrackedStress
    c = c
    kdamage = 1e-6
    F_name = E_el
    use_current_history_variable = true
    finite_strain_model = true
    uncracked_base_name = uncracked
  [../]
  [elastic_strain]
    type = ComputeMultipleInelasticStress
    inelastic_models = "trial_plasticity"
    base_name = uncracked
  []
  [hill_tensor]
    type = HillConstants
    # F G H L M N
    hill_constants = "0.01651077 0.04598923 0.06512188 0.01388889 0.01652893 0.02"
    base_name = uncracked
  []
  [trial_plasticity]
    type = HillPlasticityStressUpdate
    hardening_constant = 10
    yield_stress = 50
    base_name = uncracked
  []
  [./pfbulkmat]
    type = GenericConstantMaterial
    prop_names = 'gc_prop l visco'
    prop_values = '1e-3 0.05 5e-3'
  [../]
  [./elasticity_tensor]
    type = ComputeElasticityTensor
    C_ijkl = '120.0 80.0'
    fill_method = symmetric_isotropic
    base_name = uncracked
  [../]
[]

[Kernels]
  [./solid_x]
    type = PhaseFieldFractureMechanicsOffDiag
    variable = disp_x
    component = 0
    c = c
    use_displaced_mesh = true
  [../]
  [./solid_y]
    type = PhaseFieldFractureMechanicsOffDiag
    variable = disp_y
    component = 1
    c = c
    use_displaced_mesh = true
  [../]
  [./solid_z]
    type = PhaseFieldFractureMechanicsOffDiag
    variable = disp_z
    component = 2
    c = c
    use_displaced_mesh = true
  [../]
[]

[Bounds]
  [./irreversibility]
    type = VariableOldValueBoundsAux
    variable = bounds_dummy
    bounded_variable = c
    bound_type = lower
  [../]
  [./upper]
    type = ConstantBoundsAux
    variable = bounds_dummy
    bounded_variable = c
    bound_type = upper
    bound_value = 1.0
  [../]
[]

[Executioner]
  type = Transient
  petsc_options_iname = '-pc_type -ksp_type -ksp_gmres_restart -snes_type'
  petsc_options_value = ' lu       gmres     200 vinewtonrsls'
  solve_type = NEWTON
  dt = 5e-5
[]


[Outputs]
  exodus = true
  interval = 2 
  perf_graph = true
[]
```

**Reply by GiudGiud** [reply]
@recuero @singhgp4321 

It does not seem that it was made to work if you hit this missing property error.
Did you look through the examples in the module to see if someone combined those two material objects already?

**Reply by recuero** [reply]
It seems you are using a base name in the action. That may be preventing other objects from identifying the  rotation increment?

**Reply by ghost** [reply]
But when I do not use the base name I get other errors :

```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'uncracked_elastic_strain', requested by 'cracked_stress' is not defined on block 0
Material property 'uncracked_stress', requested by 'cracked_stress' is not defined on block 0
Material property 'uncracked_Jacobian_mult', requested by 'cracked_stress' is not defined on block 0
Material property 'uncracked_elastic_strain', requested by 'cracked_stress_face' is not defined on block 0
Material property 'uncracked_stress', requested by 'cracked_stress_face' is not defined on block 0
Material property 'uncracked_Jacobian_mult', requested by 'cracked_stress_face' is not defined on block 0
Material property 'uncracked_elastic_strain', requested by 'cracked_stress_neighbor' is not defined on block 0
Material property 'uncracked_stress', requested by 'cracked_stress_neighbor' is not defined on block 0
Material property 'uncracked_Jacobian_mult', requested by 'cracked_stress_neighbor' is not defined on block 0
Material property 'mechanical_strain', requested by 'elastic_strain' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain' is not defined on block 0
Material property 'rotation_increment', requested by 'elastic_strain' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain' is not defined on block 0
Material property 'mechanical_strain', requested by 'elastic_strain_face' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain_face' is not defined on block 0
Material property 'rotation_increment', requested by 'elastic_strain_face' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain_face' is not defined on block 0
Material property 'mechanical_strain', requested by 'elastic_strain_neighbor' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain_neighbor' is not defined on block 0
Material property 'rotation_increment', requested by 'elastic_strain_neighbor' is not defined on block 0
Material property 'strain_increment', requested by 'elastic_strain_neighbor' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_face' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_neighbor' is not defined on block 0
```

**Reply by ghost** [reply]
I've seen the examples where ComputeCrackedStress is used with IsotropicPlasticityStressUpdate (https://github.com/idaholab/moose/blob/next/modules/combined/test/tests/phase_field_fracture/crack2d_computeCrackedStress_finitestrain_plastic.i) 

> @recuero @singhgp4321
> 
> It does not seem that it was made to work if you hit this missing property error. Did you look through the examples in the module to see if someone combined those two material objects already?

**Reply by GiudGiud** [reply]
use Debug/show_material_props = true to figure out the names of the properties in each object

once everything lines up it should work as long as the rotation_increment is indeed being defined

**Reply by ghost** [reply]
```
Active Materials:
    Subdomain:  (0)
      Material Name:       cracked_stress
      Property Names:      "E_el" "Jacobian_mult" "L" "d2Fdcdstrain" "dE_el/dc" "d^2E_el/dc^2" "dstress/dc" 
                             "hist" "kappa_op" "stress" 
      Material Name:       elastic_strain
      Property Names:      "uncracked_Jacobian_mult" "uncracked_combined_inelastic_strain" "uncracked_elastic_strain" 
                             "uncracked_material_timestep_limit" "uncracked_rotation_total" "uncracked_stress" 
                             
      Material Name:       hill_tensor
      Property Names:      "uncracked_hill_constants" "uncracked_hill_tensor" "uncracked_rotation_total_hill" 
                             
      Material Name:       trial_plasticity
      Property Names:      "uncracked_effective_plastic_strain" "uncracked_hardening_variable" "uncracked_plastic_strain" 
                             "uncracked_plastic_strain_rate" 
      Material Name:       pfbulkmat
      Property Names:      "gc_prop" "l" "visco" 
      Material Name:       elasticity_tensor
      Property Names:      "uncracked_effective_stiffness" "uncracked_elasticity_tensor" 
      Material Name:       All_strain
      Property Names:      "uncracked_deformation_gradient" "uncracked_mechanical_strain" "uncracked_rotation_increment" 
                             "uncracked_strain_increment" "uncracked_strain_rate" "uncracked_total_strain" 
                             

    Active Face Materials:
    Subdomain:  (0)
      Material Name:       cracked_stress_face
      Property Names:      "E_el" "Jacobian_mult" "L" "d2Fdcdstrain" "dE_el/dc" "d^2E_el/dc^2" "dstress/dc" 
                             "hist" "kappa_op" "stress" 
      Material Name:       elastic_strain_face
      Property Names:      "uncracked_Jacobian_mult" "uncracked_combined_inelastic_strain" "uncracked_elastic_strain" 
                             "uncracked_material_timestep_limit" "uncracked_rotation_total" "uncracked_stress" 
                             
      Material Name:       hill_tensor_face
      Property Names:      "uncracked_hill_constants" "uncracked_hill_tensor" "uncracked_rotation_total_hill" 
                             
      Material Name:       trial_plasticity_face
      Property Names:      "uncracked_effective_plastic_strain" "uncracked_hardening_variable" "uncracked_plastic_strain" 
                             "uncracked_plastic_strain_rate" 
      Material Name:       pfbulkmat_face
      Property Names:      "gc_prop" "l" "visco" 
      Material Name:       elasticity_tensor_face
      Property Names:      "uncracked_effective_stiffness" "uncracked_elasticity_tensor" 
      Material Name:       All_strain_face
      Property Names:      "uncracked_deformation_gradient" "uncracked_mechanical_strain" "uncracked_rotation_increment" 
                             "uncracked_strain_increment" "uncracked_strain_rate" "uncracked_total_strain" 
                             

Active Neighboring Materials:
    Subdomain:  (0)
      Material Name:       cracked_stress_neighbor
      Property Names:      "E_el" "Jacobian_mult" "L" "d2Fdcdstrain" "dE_el/dc" "d^2E_el/dc^2" "dstress/dc" 
                             "hist" "kappa_op" "stress" 
      Material Name:       elastic_strain_neighbor
      Property Names:      "uncracked_Jacobian_mult" "uncracked_combined_inelastic_strain" "uncracked_elastic_strain" 
                             "uncracked_material_timestep_limit" "uncracked_rotation_total" "uncracked_stress" 
                             
      Material Name:       hill_tensor_neighbor
      Property Names:      "uncracked_hill_constants" "uncracked_hill_tensor" "uncracked_rotation_total_hill" 
                             
      Material Name:       trial_plasticity_neighbor
      Property Names:      "uncracked_effective_plastic_strain" "uncracked_hardening_variable" "uncracked_plastic_strain" 
                             "uncracked_plastic_strain_rate" 
      Material Name:       pfbulkmat_neighbor
      Property Names:      "gc_prop" "l" "visco" 
      Material Name:       elasticity_tensor_neighbor
      Property Names:      "uncracked_effective_stiffness" "uncracked_elasticity_tensor" 
      Material Name:       All_strain_neighbor
      Property Names:      "uncracked_deformation_gradient" "uncracked_mechanical_strain" "uncracked_rotation_increment" 
                             "uncracked_strain_increment" "uncracked_strain_rate" "uncracked_total_strain" 
                             

Active Boundary Materials:
                           


*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'rotation_increment', requested by 'hill_tensor' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_face' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_neighbor' is not defined on block 0


application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0


*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "FEProblem".

Material property 'rotation_increment', requested by 'hill_tensor' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_face' is not defined on block 0
Material property 'rotation_increment', requested by 'hill_tensor_neighbor' is not defined on block 0


application called MPI_Abort(MPI_COMM_WORLD, 1) - process 1

```

**Reply by GiudGiud** [reply]
make the hill_tensor request 'uncracked_rotation_increment' instead of 'rotation_increment'

**Reply by ghost** [reply]
Ok thanks! It works now!!

**Comment by ghost** [comment]
So now if I want to degrade the material properties by affecting the degradation function (from CrackedStress) to the yield stress (from HillElastoPlasticity). To do that I'm "getting" the c from ComputeCrackedStress to HillElastoPlasticityStressUpdate. But now I'm getting this error 

```
*** ERROR ***
Cyclic dependency detected in object ordering:
elastic_strain <- cracked_stress <- elastic_strain
```

I don't think I understand what this means

**Reply by GiudGiud** [reply]
This means the objects (the material properties computed by each material) depend on each other in a circular fashion. We need one to compute the other, but the other needs the first one to be computed. The dependencies should not form a cycle

**Reply by ghost** [reply]
Ok so I can't couple these two together because they are both being called by ComputeMultipleInelasticStress

**Reply by GiudGiud** [reply]
This part should be fine. They can both be depended on by a 3rd material

they just cant depend both on each other

**Reply by ghost** [reply]
But I'm only trying to have HillElastoPlasticityStressUpdate depend on CrackedStress and not the other way around.

**Reply by GiudGiud** [reply]
in the code for the crackedStress you are using the elastic_strain. If you dont want that, then change that code

**Reply by ghost** [reply]
In cracked stress I want to use the elastic strain to calculate the free energy. What I want is to be able to use the damage variable ("c" from CrackedStressed) in HillElastoPlasticity to degrade the yield stress

**Reply by GiudGiud** [reply]
can you use the property from the previous time step? with getMaterialPropertyOld

this will break the circle of dependencies

**Reply by ghost** [reply]
I tried that and still got the same error

**Reply by GiudGiud** [reply]
what did you change?

**Reply by ghost** [reply]
In ComputeCrackedStress I added 
`_dam(declareProperty<Real>(_uncracked_base_name + "dam")),` 

 with `_dam[_qp] = _c[_qp];`

and in HillElastoPlasticity:
```
_dam(this->template getGenericMaterialProperty<Real, is_ad>(this->_base_name + "dam")),
_dam_old(this->template getMaterialPropertyOld<Real>(this->_base_name + "dam")),
```
with the new yield stress:

```
      _yield_stress  =   _yield_stress * ( 1 - _dam_old[_qp]) * ( 1 - _dam_old[_qp]);

```

**Reply by GiudGiud** [reply]
> _dam(this->template getGenericMaterialProperty<Real, is_ad>(this->_base_name + "dam")),
_dam_old(this->template getMaterialPropertyOld<Real>(this->_base_name + "dam")),

you are still retrieving the current value. So the dependency is still there. If you want to break the dependency you have to ONLY retrieve the old value

**Reply by ghost** [reply]
Oh, ok! Thank you !!

**Comment by ghost** [comment]
Hi, another question concerning HillElastoPlasticityStressUpdate, the eq_plastic_strain_inc doesn't return the usual equivalent plastic strain? 
![image](https://github.com/idaholab/moose/assets/157137189/55bf4a8f-3cbd-4a04-ac06-39d15b23c59b)
as it's also defined and called in FiniteStrainPlasticMaterial

**Reply by GiudGiud** [reply]
`_inc` stands for increment. So it's not the full value of strain, only the difference over a time step

**Reply by ghost** [reply]
`  _effective_inelastic_strain[_qp] = _effective_inelastic_strain_old[_qp] + eq_plastic_strain_inc;`
But even when accumulated it doesn't return the same values as the equivalent plastic strain

**Reply by GiudGiud** [reply]
@gambka

**Reply by gambka** [reply]
@singhgp4321 is better suited to answer this question.

**Reply by recuero** [reply]
The definition of the effective plastic strain might be different from other objects. There is a report and references on this set of anisotropic objects. 

I would look in the references and change things if one wants to have the same definition.

**Reply by ghost** [reply]
Thanks all, I added it manually :+1:

**Comment by ghost** [comment]
Hello,
My main objective is to integrate damage into Hill anisotropic plasticity. So first, I attempted to use HillElastoPlasticityStressUpdate in combination with ComputeCrackedStress and ComputeMultipleInelasticStress. My approach involved degrading the yield stress in HillElastoPlasticityStressUpdate using the damage_old parameter from ComputeCrackedStress. However, this resulted in an abnormal increase in plastic strain once the yield stress was reached. In another attempt, I incorporated the phase field formulation into ComputeMultipleInelasticStress by defining the degraded stress and its derivatives, along with the free energy and its derivatives (dstress_dc, F, dFdc, d2Fdc2, d2Fdcdstrain). Unfortunately, this also produced incorrect results, showing the same abnormal surge in plastic strain. 
I suspect the error originates from ComputeMultipleInelasticStress, but I am unsure where to look exactly. Do you have any suggestions? Thanks!

**Reply by singhgp4321** [reply]
@nourshabib I am trying to understand the specifics of your damage model and how you are trying to use it alongside HillElastoPlasticity. Is there a document on your damage model that you could point me to?

There can be several reasons for getting unusually high plastic strain. Is your damage parameter showing reasonable values, same thing about the yield stress after the onset of damage? I will first ensure that the damage parameter is not jumping to an abruptly high value, or the yield stress is not dropping down too fast. If that is not the case then I will need to look into the code that everything looks fine there.

**Reply by ghost** [reply]
Phase field for fracture in ductile material so I want to include a plastic free energy term in the free energy formulation (specifically the 'yield stress * the equivalent plastic strain * the degradation function' in the free energy formula ) and degrade the yield stress (degradation function * yield_stress), as in "A phase-field formulation for fracture in ductile materials: Finite deformation balance law derivation, plastic degradation, and stress triaxiality effects": https://www.sciencedirect.com/science/article/pii/S0045782516311069, haven't seen this done with Hill plasticity, but it shouldn't be  that different from the von mises plasticity. The unusually high plastic strain thus leads to the high damage values and the sharp decrease in stress. 
What I've done:
- In HillElastoPlasticityStressUpdate I added the degradation function (1-c)^2 and affected it to the yield stress : yield_stress * (1-c)^2
- In ComputeCrackedStress I added the plastic free energy term to the free energy and its derivatives as : yield_stress *  eqv_plastic_strain * (1-c)^2

And this lead to high plastic strain, sharp increase in damage and sharp stress decrease.
